### PR TITLE
feat(async-flow): `startGeneration`: a new synthetic `LogEntry`

### LINKED ISSUE #9384: async-flow: need incarnation number in the log classifying entries
Another thing that must happen before we commit to durable log state.

### PR DESCRIPTION
**Pull request by michaelfig**
closes: #9384

## Description

This pull request introduces a `startGeneration` log entry to capture the generation (restart) number for subsequent log entries. The most important changes are grouped below by theme.

### Log Entry Enhancements:

* Introduced `startGeneration` and updated `LogEntryShape` to include a new log entry type for starting a generation (`packages/async-flow/src/type-guards.js`).

### Log Store Improvements:

* Retrieve the current generation number from a `zone.mapStore` and use it in LogStore ephemera to track its generation (`packages/async-flow/src/log-store.js`).
* Modified `prepareLogStore` to define a predicate for filtering log entries, and added methods for retrieving unfiltered log entries (`packages/async-flow/src/log-store.js`).
* Updated test cases to validate the new generation number and unfiltered log entries (`packages/async-flow/test/log-store.test.js`).

### Replay Membrane Adjustments:

* Added support for `startGeneration` in the replay membrane to handle generation-specific operations (`packages/async-flow/src/replay-membrane.js`).

### Type Definitions:

* Updated `LogEntry` and `FutureLogEntry` types to include the new generation entry type (`packages/async-flow/src/types.ts`).

### Security Considerations

n/a

### Scaling Considerations

Low overhead implementation.

### Documentation Considerations

Not necessary yet, just laying groundwork for the infrastructure so that a future change can use it to implement asyncFlow versioning.

### Testing Considerations

Some new `*Unfiltered` methods of the `LogStore` now have some tests.

### Upgrade Considerations

Designed to enable future upgrade of async flows.

### CODE REVIEW

File: packages/async-flow/src/log-store.js
Comment (erights) [inline packages/async-flow/src/log-store.js]: Seems strange to make this parameterizable rather than just calling a fixed local filter function. I notice this parameter is not yet used. Is there a motivating or expected use?

File: packages/async-flow/src/log-store.js
Comment (erights) [inline packages/async-flow/src/log-store.js:158]: It is so surprising for a method named `peekSomething` to have observable side effects that this should probably now be renamed. It should at least have a clear note in any relevant place about what side effect it now does have.

File: packages/async-flow/src/log-store.js
Comment (erights) [inline packages/async-flow/src/log-store.js]: I'm surprised this is an `if` and the one on line 160 above is a `while`. If a `while` is needed there, wouldn't it be needed here as well? Likewise vice versa.

File: packages/async-flow/src/log-store.js
Comment (erights) [reply packages/async-flow/src/log-store.js:158]: Even better would be to revise this function so that it was again a pure query function.

File: packages/async-flow/src/log-store.js
Comment (erights) [inline packages/async-flow/src/log-store.js]: first `pushEntry` per what? I would think this need to be per logStore instance. But this seems to be per logStore preparation. I don't see how this can work, or even pass any test with multiple activations (instances) of one asyncFlow (class).

File: packages/async-flow/src/log-store.js
Comment (erights) [inline packages/async-flow/src/log-store.js]: See above for my confusion about the scope of `prependedEntries`.

File: packages/async-flow/src/log-store.js
Comment (erights) [inline packages/async-flow/src/log-store.js]: cute ;)

File: packages/async-flow/src/replay-membrane.js
Comment (erights) [inline packages/async-flow/src/replay-membrane.js]: These lines are surprisingly asymmetric. How about just 
```suggestion
    startGeneration,
    doFulfill,
    doReject,
    // doCall, // unimplemented in the current plan
```
? 

Alternatively, change the others to follow your first line convention, which would at least be consistent. I prefer the briefer form though.

File: packages/async-flow/src/type-guards.js
Comment (erights) [inline packages/async-flow/src/type-guards.js]: Why is it that this one needs the name indirection but the other ops do not? In general my reaction is that I understand why such indirection adds value to C programs, but not to JS programs. Especially not to JS programs that are already checked by TS types that can include such string constants in the type.

But the main thing is consistency. Do this for either all or none.

File: packages/async-flow/src/type-guards.js
Comment (erights) [inline packages/async-flow/src/type-guards.js]: What is gained from the extra indirection here of first forming the list of alternatives and then feeding them to an `M.or`? Do you expect to make some other use of `LogEntryAlternatives`?

**Review by erights — CHANGES_REQUESTED** [review]

File: packages/async-flow/src/async-flow.js
Comment (mhofman) [inline packages/async-flow/src/async-flow.js]: @warner I'd like to align this with `incarnationNum`. Is that 0 or 1 based?

File: packages/async-flow/src/log-store.js
Comment (michaelfig) [reply packages/async-flow/src/log-store.js:158]: Thanks for pointing this out.  Purified again.

File: packages/async-flow/src/log-store.js
Comment (michaelfig) [reply packages/async-flow/src/log-store.js]: No motivating use.  Made into a fixed local filter function as suggested.

File: packages/async-flow/src/log-store.js
Comment (michaelfig) [reply packages/async-flow/src/log-store.js]: It needs to be first pushEntry per logStore instance.  Made it so.

File: packages/async-flow/src/type-guards.js
Comment (michaelfig) [reply packages/async-flow/src/type-guards.js]: Removed the indirection.

File: packages/async-flow/src/type-guards.js
Comment (michaelfig) [reply packages/async-flow/src/type-guards.js]: It was a relic from the generation-per-LogEntry approach.  Simplified.

File: packages/async-flow/src/replay-membrane.js
Comment (michaelfig) [reply packages/async-flow/src/replay-membrane.js]: Made briefer.

File: packages/async-flow/src/type-guards.js
Comment (mhofman) [reply packages/async-flow/src/type-guards.js]: In general I do like constants to help refactoring, but in this case these should never change, so I'm enticed to agree that we don't need a const **if** our typing is correct.

File: packages/async-flow/src/log-store.js
Comment (michaelfig) [reply packages/async-flow/src/log-store.js]: Factored differently now, but the reason it worked was that the `while` was in `self.peekEntry()`, called from `nextEntry` line 173, so `nextEntry` only needed to know if the final result passed the predicate.

File: packages/async-flow/src/log-store.js
Comment (erights) [inline packages/async-flow/src/log-store.js]: Given your method naming convention `foo` vs `unfilteredFoo`, should these state variables be renamed to

```suggestion
      index: 0,
      unfilteredIndex: 0,
```
?

File: packages/async-flow/src/log-store.js
Comment (erights) [reply packages/async-flow/src/log-store.js]: This still doesn't seem quite right. Your code does indeed do per instance, which is indeed what I suggested. But don't we need per instance per incarnation? AFAICT, this only gets set by the `init` function, which is only per instance, not per incarnation.

File: packages/async-flow/src/log-store.js
Comment (mhofman) [inline packages/async-flow/src/log-store.js]: Why are `entriesToPrepend` in the state of the log store (which as far as I know has a 1:1 mapping with the flow invocation.

@erights correct me if I'm wrong but the ephemera would be the place to keep this, right? It would mean that any restart of the flow that pushes at least one entry would insert a start generation entry first, even if the restart happens in the same generation, but I think that's a correct behavior.

File: packages/async-flow/src/log-store.js
Comment (mhofman) [inline packages/async-flow/src/log-store.js]: Technically the `self.isReplaying()` is redundant since a `startGeneration` entry would always appear before another entry in the log.

I suppose we may have other invisible entries in the future that are at the tail of the log, but I think that would mess other cases of "is replaying check", so that may be something we likely want to prevent anyways.

```suggestion
        while (!entryIsVisible(result)) {
          self.isReplaying() || Fail`Unexpected entry at log tail: ${result}`;
          result = self.nextUnfilteredEntry();
        }
```

**Review by mhofman — COMMENTED** [review]
I think I'd like to see a test of both a log reset and a flow restart that shows the generation of an invocation getting bumped when inserting new log entries. This likely can be done using explicit reset/restarts, and without a simulated liveslots upgrade (especially because I haven't landed the liveslots test framework improvements yet, and we don't have testing hooks to peek into the state of async flow, which would possibly be covered by #9753)

File: packages/async-flow/src/log-store.js
Comment (erights) [reply packages/async-flow/src/log-store.js]: Yes, ephemera would be correct.

File: packages/async-flow/src/async-flow.js
Comment (michaelfig) [reply packages/async-flow/src/async-flow.js]: `incarnationNumber` is 0-based.  I will make `logGeneration` start the same.

File: packages/async-flow/src/log-store.js
Comment (michaelfig) [reply packages/async-flow/src/log-store.js]: I think they should, so I've renamed the state variables accordingly.

File: packages/async-flow/src/log-store.js
Comment (michaelfig) [reply packages/async-flow/src/log-store.js]: That's what I converged on, as well.

File: packages/async-flow/src/log-store.js
Comment (michaelfig) [reply packages/async-flow/src/log-store.js]: Committed your suggestion.

File: packages/async-flow/src/log-store.js
Comment (erights) [reply packages/async-flow/src/log-store.js]: IIUC, looks good now. Thx.

**Review by erights — APPROVED** [review]
All my issues are dealt with well, so approving to remove my blockage. But you should still iterate with @mhofman to his satisfaction before merging.

**Review by mhofman — COMMENTED** [review]
LGTM (once the test snapshots are updated to reflect the new 0 based generation)

**Review by mhofman — APPROVED** [review]

**Comment by mhofman** [comment]
I am wondering if we should include a `lastGenerationPush` or something like that in the state of the log store, and some assertion that the start generation never regresses (it can stay the same).

**Comment by michaelfig** [comment]
> I am wondering if we should include a `lastGenerationPush` or something like that in the state of the log store, and some assertion that the start generation never regresses (it can stay the same).

I appreciate the need to have guarantees like this, but adding extra code for validation didn't strike me as the best thing to do.  Ruminating on it, I'd prefer to encapsulate the generation number within `prepareLogStore` (where its increment-only behaviour would be easily verifiable by inspection) rather than passing it in from outside.

**Comment by mhofman** [comment]
> I'd prefer to encapsulate the generation number within `prepareLogStore` (where its increment-only behaviour would be easily verifiable by inspection) rather than passing it in from outside.

Yeah I think that's what was making me uneasy, that async-flow could mistakenly not pass the right generation. I prefer it container like this much better.

**Pull request merged**
